### PR TITLE
feat(UI builder): Add VPC cursor simulation and new roles and element types for the screen reader narration computation, 

### PR DESCRIPTION
**Pull request by adamsamec**
#### Description of changes

This PR introduces the simulated virtual PC (VPC) cursor navigation mode, also known as browse or scan mode, to the screen reader narration computation feature of the UI builder. This mode can be toggled in the "Use" mode via the "Enable virtual cursor" checkbox at the top of the page, and starts working when the keyboard focus is in the canvas.

The following keyboard commands work when the virtual cursor is enabled:

* Control + "." (dot): Moves the virtual cursor to the next element.
* Control + "," (comma): Moves the virtual cursor to the previous element.
* Enter: Clicks the element focused with the virtual cursor.
* Shift + F10: Opens the context menu for the element focused with the virtual cursor.

Another change in this PR is so that, in both the standard and the new virtual cursor mode, the computed screen reader narration recognises also the following new element types and ARIA roles:

* role="link"
* &lt;input type="search"&gt; and role="searchbox"
* role="textbox"
* role="treeitem"

It also adds the support for the following new group and composite roles:

* role="menubar"
* role="treegrid"

When the virtual cursor mode is enabled, these element types, roles and CSS selectors are recognised on the top of the element types and roles already supported by the standard mode:

* h1, h2, h3, h4, h5, h6
* role="listbox", role="listitem"
* .ui-text:not(.ui-checkbox__label)

In addition, this PR introduces some bug fixes and other improvements. The main improvement being the platform definitions inheritance is now possible. That is, now a new platform doesn't have to be defined from scratch, and can instead be defined based on another platform, defining only what has changed. Currently, there are two platforms: "Win/JaWS" and "Win/JAWS/VPC", where "Win/JAWS/VPC" inherits definitions from "Win/JAWS".

When the virtual cursor mode is enabled and the keyboard commands mentioned above are used to navigate inside the canvas, the "Win/JAWS/VPC" platform is used, otherwise the "Win/JAWS" platform is used.

### CODE REVIEW

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 1]
```diff
@@ -62,6 +64,10 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
     [],
   );
 
+  const [focusableElements, setFocusableElements] = React.useState([]);
```
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/fluentui/react-builder/src/components/Canvas.tsx]: Outdated

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 2]
```diff
@@ -96,6 +102,43 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
     [onMouseUp, hideDropSelector],
   );
 
+  const handleKeyDown = React.useCallback(
+    e => {
+      switch (e.keyCode) {
```
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/fluentui/react-builder/src/components/Canvas.tsx]: Outdated

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 3]
```diff
@@ -96,6 +102,43 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
...
+        case 40:
+        case 38:
+          focusableElements[currentIndex].classList.remove('virtual-focused');
+          setIndex(idx => {
+            const modifier = e.keyCode === 40 ? 1 : -1;
+            const nextIndex = idx + modifier;
+            // if nextIndex is bigger than number of elements move it to 0
+            // if nextIndex is smaller than 0 move it to the lastElement
+            // otherwise move to the nextIndex
```
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/fluentui/react-builder/src/components/Canvas.tsx]: Outdated

File: packages/fluentui/react-builder/src/components/ReaderNarration.tsx [hunk 4]
```diff
@@ -42,21 +44,23 @@ export const ReaderNarration: React.FunctionComponent<ReaderNarrationProps> = ({
...
   // Recomputes the narration text upon every render.
+  // eslint-disable-next-line react-hooks/exhaustive-deps
   React.useEffect(() => {
     if (!ref.current || aomMissing) {
       return;
     }
 
     // Recompute and save the focusable elements and their paths for the tree rooted at the selector's element upon every selector change
-    if (prevSelector !== selector) {
+    if (selector !== prevSelector) {
```
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/fluentui/react-builder/src/components/ReaderNarration.tsx:56]: Outdated

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 5]
```diff
@@ -96,6 +102,43 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
...
+            return newIndex;
+          });
+          break;
+        case 13:
+          focusableElements[currentIndex].click();
+          return;
+        case 121:
+          if (e.shiftKey) {
+            const eve = document.createEvent('MouseEvents');
+            eve.initMouseEvent('contextmenu', true, true, window, 1, 0, 0, 0, 0, false, false, false, false, 2, null);
```
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/fluentui/react-builder/src/components/Canvas.tsx]: outdated

File: packages/fluentui/react-builder/src/narration/NarrationComputer.tsx [hunk 6]
```diff
@@ -305,14 +306,15 @@ export class NarrationComputer {
   // Computes and stores the accessible description part of the narration for the given definitionName, element and platform.
   computeDescription(definitionName: string, element: HTMLElement, platform: string) {
     // First, handle some special case conditions
+    // Note: Actually, this special case condition is currently not in use by any platform, but we are keeping it here for a potential use in future
```
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/fluentui/react-builder/src/narration/NarrationComputer.tsx]: outdated

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 7]
```diff
@@ -62,6 +64,10 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
     [],
   );
 
+  const [focusableElements, setFocusableElements] = React.useState([]);
+  const [currentIndex, setIndex] = React.useState(0);
+  const [currentFocusedNode, setCurrentFocusedNode] = React.useState(null);
```
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/fluentui/react-builder/src/components/Canvas.tsx]: outdated

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 8]
```diff
@@ -96,6 +102,43 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
...
 
+  const handleKeyDown = React.useCallback(
+    e => {
+      switch (e.keyCode) {
+        case 40:
+        case 38:
+          focusableElements[currentIndex].classList.remove('virtual-focused');
+          setIndex(idx => {
+            const modifier = e.keyCode === 40 ? 1 : -1;
+            const nextIndex = idx + modifier;
```
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/fluentui/react-builder/src/components/Canvas.tsx]: outdated

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 9]
```diff
@@ -66,6 +68,10 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
     [],
   );
 
+  const [virtualCursorElements, setVirtualCursorElements] = React.useState([]);
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:71]: nit: please use a type here (might not be correct in the suggestion)
```suggestion
  const [virtualCursorElements, setVirtualCursorElements] = React.useState<HTMLElement>([]);
```

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 10]
```diff
@@ -66,6 +68,10 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
...
+  const [vcIndex, setVcIndex] = React.useState(0);
+  const [focusedVcElement, setFocusedVcElement] = React.useState(null);
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:73]: nit: use a type here
```suggestion
  const [focusedVcElement, setFocusedVcElement] = React.useState<HTMLElement>(null);
```

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 11]
```diff
@@ -100,6 +106,69 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
     [onMouseUp, hideDropSelector],
   );
 
+  const handleKeyDown = React.useCallback(
+    event => {
+      switch (
+        event.keyCode // Begin switch 1
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:112]:
```suggestion
        event.key // Begin switch 1
```

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 12]
```diff
@@ -100,6 +106,69 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
...
+      ) {
+        case 190: // "." (dot) key - Moves to the next element
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:114]:
```suggestion
        case 'period': // "." (dot) key - Moves to the next element
```

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 13]
```diff
@@ -100,6 +106,69 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
...
+        case 188: // "," (comma) key - Moves to the previous element
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:115]:
```suggestion
        case 'comma': // "," (comma) key - Moves to the previous element
```

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 14]
```diff
@@ -100,6 +106,69 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
...
+          // } // End if 1
+          return;
+        case 13: // Enter key - Clicks the current element
+          virtualCursorElements[vcIndex].click();
+          return;
+        case 121: // F10 key
+          if (event.shiftKey) {
+            // Begin if 1
+            const mouseEvent = document.createEvent('MouseEvents');
+            mouseEvent.initMouseEvent(
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:145]: `initMouseEvent` is deprecated please use the constructor instead `new MouseEvent()`

More info here: https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/MouseEvent

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 15]
```diff
@@ -153,6 +222,66 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
...
+              '[role="checkbox"]',
+              // '[role="dialog"]',
+              '[role="gridcell"]',
+              '[role="link"]',
+              // '[role="list"]',
+              '[role="listbox"]',
+              '[role="listitem"]',
+              '[role="log"]',
+              // '[role="menu"]',
+              // '[role="menubar"]',
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:256]: Are you intentionally leaving the commented roles in the codebase ?

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 16]
```diff
@@ -100,6 +106,69 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
...
+                ? virtualCursorElements.length - 1
+                : nextIndex;
+
+            virtualCursorElements[newIndex].classList.add('virtual-focused');
+            setFocusedVcElement(virtualCursorElements[newIndex]);
+            return newIndex;
+          }); // End setVcIndex
+          // } // End if 1
+          return;
+        case 13: // Enter key - Clicks the current element
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:138]:
```suggestion
        case 'Enter': // Enter key - Clicks the current element
```

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 17]
```diff
@@ -100,6 +106,69 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
...
+          virtualCursorElements[vcIndex].click();
+          return;
+        case 121: // F10 key
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:141]:
```suggestion
        case 'F10': // F10 key
```

File: packages/fluentui/react-builder/src/narration/NarrationComputer.tsx [hunk 18]
```diff
@@ -209,85 +256,139 @@ export class NarrationComputer {
...
-  // Returns the definition name based on the given element, platform and definition type.
-  getDefinitionName(element: IAriaElement, platform: string, definitionType: string): string {
-    // Determine the definitions source by the definition type
-    const definitionTypes: Record<string, any> = {
-      readingOrder: SRNC.readingOrder[platform],
-      stateRules: SRNC.stateRules[platform],
-    };
-    const definitions = definitionTypes[definitionType];
+  // Returns the platform inheritance list for the given platform
+  getPlatformInheritance(platform: string): string[] {
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/narration/NarrationComputer.tsx:260]: Does it make sense to create a type for `platform` to constrain the strings that are allowed ?

File: packages/fluentui/react-builder/src/components/ReaderNarration.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/ReaderNarration.tsx:56]: I don't remember  :-) But it does not matter, does it?

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 9 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/Canvas.tsx:71]: Okay, fixed.

File: packages/fluentui/react-builder/src/components/ReaderNarration.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/fluentui/react-builder/src/components/ReaderNarration.tsx:56]: I think it's always good to not add unnecessary changes to the changelist, makes reviews easier, but not a big deal I agree

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 10 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/Canvas.tsx:73]: Fixed as well.

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 11 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/Canvas.tsx:112]: Used event.code instead to be independent of different keyboard layouts.

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 12 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/Canvas.tsx:114]: Fixed.

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 13 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/Canvas.tsx:115]: Applied

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 15 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/Canvas.tsx:256]: Yes, I am. The commented selectors are intentionally not included because they are narrated as landmarks or groups when enterring elements. I will add a comment explaining why they are commented out.

Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/Canvas.tsx:256]: I wanted to keep them there in case they will be added in future and so tthat it's clear that these selectors are not missed out but ARE intentionally not included.

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 17 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/Canvas.tsx]: Applied

File: packages/fluentui/react-builder/src/narration/NarrationComputer.tsx (hunk 18 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/narration/NarrationComputer.tsx:260]: Well, not sure what benefit it will bring.

Comment (ling1726) [reply packages/fluentui/react-builder/src/narration/NarrationComputer.tsx:260]: easier maintenance/readability IMO

File: packages/fluentui/react-builder/src/components/ReaderNarration.tsx [hunk 19]
```diff
@@ -66,7 +72,7 @@ export const ReaderNarration: React.FunctionComponent<ReaderNarrationProps> = ({
           // Preselect the path and the narration element to the first focusable element with tabindex >= 0
           const preselectedElementItem =
             focusableElementsItems.find(item => item.element.tabIndex >= 0) || focusableElementsItems[0];
-          selectedElementPath = preselectedElementItem.path;
+          selectedElementPath = preselectedElementItem.path.join(pathSeparator);
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/components/ReaderNarration.tsx:75]: Why is this change necessary ? was it not a string before ? 

Also what is the current type ? seems to me that it is `any` which makes reviewing this harder

File: packages/fluentui/react-builder/src/components/ReaderNarration.tsx [hunk 20]
```diff
@@ -4,18 +4,21 @@ import { NarrationComputer, IAriaElement } from './../narration/NarrationCompute
 
 const computer = new NarrationComputer();
 let prevSelector = null;
+let prevVcElement = null;
+let allowVirtualCursor = true;
 let focusableElements = {};
 let elementsPaths = [];
 let selectedElementPath = null;
 let prevNarrationElement = null;
 const aomMissing = !window.hasOwnProperty('getComputedAccessibleNode');
```
Comment (ling1726) [inline packages/fluentui/react-builder/src/components/ReaderNarration.tsx:13]: You don't have to do this PR, but please make sure that you create another PR to make these variables refs in the component with the `useRef` hook https://reactjs.org/docs/hooks-reference.html#useref

File: packages/fluentui/react-builder/src/narration/NarrationComputer.tsx (hunk 18 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/narration/NarrationComputer.tsx]: Okay, changed.

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 14 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/Canvas.tsx]: Actually, the current code for showing the context menu does not work. So resolved it for now by removeing the code. The possibility to show a context menu on an element focused with the virtual cursor is not essential. Anyway, we did some investigation with Milan and couldn't find a way how to dispatch an event which would open a context menu. [Here is a JSFiddle](https://jsfiddle.net/kolaps33/6jtufpxq/49/) which we used for testing.

File: packages/fluentui/react-builder/src/components/ReaderNarration.tsx (hunk 20 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/ReaderNarration.tsx:13]: Good point, so lets leave it for another PR in the future.

File: packages/fluentui/react-builder/src/components/ReaderNarration.tsx (hunk 19 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/ReaderNarration.tsx:75]: Okay, I  have added types for all the global variables in this file.

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 21]
```diff
@@ -100,6 +106,42 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
...
   );
 
+  const handleKeyDown = React.useCallback(
+    event => {
+      switch (
+        event.code // Begin switch 1
+      ) {
+        case 'Period': // Moves to the next element
+        case 'Comma': // Moves to the previous element
+          // if (event.ctrlKey) { // Begin if 1
```
Comment (miroslavstastny) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:116]: With this line commented out, you can navigate using just comma and period. That does not match the PR description.
Also what is the intentional hotkey on Mac? Usually whatever is *Control* on Windows we do *Command* on Mac. But *Command,* is already used on Mac.
I am ok with using *Control* regardless the platform.

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 22]
```diff
@@ -336,8 +445,12 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
               )}
               {inUseMode && <EventListener capture type="focus" listener={handleFocus} target={document} />}
               {renderJSONTreeToJSXElement(jsonTree, renderJSONTreeElement)}
+              {inUseMode && enabledVirtualCursor && (
+                <EventListener type="keydown" listener={handleKeyDown} target={document} />
+              )}
               <div style={{ bottom: '0', position: 'absolute' }}>
                 <ReaderNarration
```
Comment (miroslavstastny) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:452]: Virtual cursor navigates even to `ReaderNarration`. Is it is not part of the designed UI, it is not expected.

File: packages/fluentui/react-builder/src/components/Canvas.tsx [hunk 23]
```diff
@@ -238,6 +341,12 @@ export const Canvas: React.FunctionComponent<CanvasProps> = ({
             outline-offset: -1px;
           }
           `,
+        inUseMode &&
```
Comment (miroslavstastny) [inline packages/fluentui/react-builder/src/components/Canvas.tsx:344]: When I use the virtual cursor and later disable it but stay in the use mode, the virtual cursor frame remains rendered. I think it should be removed.

File: packages/fluentui/react-builder/src/components/Designer.tsx [hunk 24]
```diff
@@ -221,6 +223,10 @@ const stateReducer: Reducer<DesignerState, DesignerAction> = (draftState, action
       }
       break;
 
+    case 'ENABLE_VIRTUAL_CURSOR':
+      draftState.enabledVirtualCursor = action.enabledVirtualCursor;
```
Comment (miroslavstastny) [inline packages/fluentui/react-builder/src/components/Designer.tsx:227]: I would expect once I enable the VC, first item in the design is selected. Both when enabled it for the first time as well as when disabled and enabled again.

**Review by miroslavstastny — COMMENTED** [review]
The VPC does not work for a `Dialog` - regular virtual cursor goes into the dialog and traps focus there. The simulated cursor stays outside, navigates only to Dialog's header, not to the button and is not trapped. Also when the `Dialog` is closed while the simulated VC is inside, the VC gets lost.

Link to repro on localhost:
http://localhost:8080/builder#tree_lz=N4IgLgngDgpiBcIAmBLAbiANMlBnKANgIYQByRAtnIqhtgK70pIIgBGTBSMATgLQ8A9oLBYQAEnGRYggGasAyhAptBBABQ8YRAMZgAdDAIwqAOzABKMVCFRcCUAGsYEVhxRdeA4aOw6AFh5IWqYIANqg0tQgAEL0YGCCodio%2BMRklNFxCUlijMysEGiCAJwAHABMUADssjDy2JJRcorKqhpaugZGJjDmVtg2gnYOIDpJYH2iiNmJySAo46HwkdDRAMJEBAQAkktiqYQk5FSsm9t7uY1Say2ISipqmtp6hsZmlta29isAvv%2B-TCrWBueJzA54I4ZU6IDg5eb5FiIaq4AAMdQAzKjUfRHA0JDcZPJ7m0np1Xj0PgMQEMRisxoEuCFwiAAB4gAC6-yB4DWrAAIigtoIAOYQtLHTICoUEUV5JhIkAANgArAAWADuGtMjjV1RFRDETVuxJAD3azy6b16-S%2Bwx%2BoCWk3MoyirFmVxwEuhWTBnsRrH8zhgFQIACt6DxqkbCTA7mbSR0Xt13lNqbSHSBnK5EEGYCHw5Ho34JlN3X7TAACFCmXDMGCV1DCsXYRa5eluxDnXb7FKQ9InDZbHue41E1qPJNWylpu104DckB1gBe0VwFGHMFwokX-m03B4rAAEvveGIwDwUCKRWeO3yZhXxVDB%2BX4THmqbzWTk9aqXPMwEQTMvAYQgAA8rAVaCs2nIAr8XK-EAA

**Comment by miroslavstastny** [comment]
@jurokapsiar - it would be great if @kolaps33 could test this.

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 21 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/Canvas.tsx]: Sorry, the line should not be commented out. Control + comma and Control + dot should be used for navigation regardless of the platform, as mentioned in the PR description.

File: packages/fluentui/react-builder/src/components/Canvas.tsx (hunk 22 above)
Comment (adamsamec) [reply packages/fluentui/react-builder/src/components/Canvas.tsx:453]: Agreed via chat to leave it as it is.

**Review by jurokapsiar — APPROVED** [review]

File: packages/fluentui/react-builder/src/components/Designer.tsx (hunk 24 above)
Comment (jurokapsiar) [reply packages/fluentui/react-builder/src/components/Designer.tsx]: fixed - focus goes to iframe and keyboard shortcuts start to work. also, tooltip with hint is displayed
